src/compile.rs

```diff
@@ -475,6 +476,40 @@ pub fn validate_rules(statements: &BTreeSet<ValidatedStatement>) -> Result<(), C
     errors.into_result(())
 }
 
+pub fn prevalidate_functions(
+    functions: &mut FunctionMap,
+    types: &TypeMap,
+) -> Result<(), CascadeErrors> {
+    initialize_castable(functions, types);
+    // We initialize to 1 just to let the loop start once
+    let mut num_changed: u64 = 1;
+    while num_changed > 0 {
+        num_changed = determine_castable(functions, types);
+    }
+
+    let (mut terminated_functions, mut nonterm_functions) = initialize_terminated(functions);
+
+    if terminated_functions.is_empty() && !nonterm_functions.is_empty() {
+        let mut error: Option<CompileError> = None;
+
+        for func in functions.values() {
```

Comment:
You are correct about search_for_recursion catching the error so we don't need this here.

The one thing we don't know here or in search_for_recursion is the nature of the recursive loop.  There may be more than one loop.  We just know that there is a list of functions that did not terminate.

Reply:
I updated the code to work the way you outlined, and things are working.  But I just want confirmation that our logic is correct.

Reply:
> I updated the code to work the way you outlined, and things are working. But I just want confirmation that our logic is correct.

Was on the wrong comment sorry.  That was for the:
> Isn't that the case that if we do find a single terminating thing we are ourselves terminating? If we find a non-terminating thing there still may be other terminating things to find. Or do I have my logic just 100% backwards on that?

comment instead.

Reply:
And to clarify something here.  We are not overwriting the error. `add_or_create_compile_error` will either create an error if the error passed in is `None` or it will call `.add_additional_message()`.  The output we will currently get looks something like: 
```
error: Recursive Function call found
   ┌─ data/error_policies/functions_recursion.cas:7:5
   │
 7 │     fn other_read(domain source) {
   │        ^^^^^^^^^^ Calls cannot recursively call each other
   │
   ┌─ data/error_policies/functions_recursion.cas:2:5
   │
 2 │     fn read(domain source) {
   │        ^^^^ Calls cannot recursively call each other
   │
   ┌─ data/error_policies/functions_recursion.cas:12:5
   │
12 │     fn third_read(domain source) {
   │        ^^^^^^^^^^ Calls cannot recursively call each other

Error: Custom { kind: InvalidData, error: "Invalid policy" }
```

Reply:
OH!  This makes so much more sense now.

Okay, this is fine for now then.  In the future, we could probably enhance the error message to be more helpful by explicitly identifying one or more loops, but this seems good enough for our purposes now.

Reply:
Based on that, I'll go ahead and give this another full read-over after lunch.  It seems like all the outstanding issues are addressed, so hopefully it'll all look good at that point.